### DISCUSSION TITLE
How to run for_windows commands in i3ipc?

### DISCUSSION THREAD

**Discussion by medwatt** [i3/i3#6251] [Q&A]
I have the following python code that launches applications from a dictionary. 

```python
applications = {
    "btop": {
        "command": "kitty --single-instance --title 'btop' --class kitty_btop btop",
        "properties": [
            "floating enable",
            "resize set 800 800",
            "move position center"
        ]
    },
}

def launch_app(self, app_name):
    if app_name in applications.keys():
        command = applications[app_name]["command"]
        properties = applications[app_name]["properties"]

        run_command(self.i3.command, f"exec {command}")

        if properties:
            def on_window_event(i3, event):
                window_id = event.container.window
                command = f"[id={window_id}] " + ", ".join(properties)
                run_command(self.i3.command, command)
                self.i3.off(on_window_event)
            self.i3.on(Event.WINDOW_NEW, on_window_event)
```

This code works just as expected. However, because I have to wait until the window is created before the properties are applied, there's a noticeable delay during which the application is switched from tiled to floating.

I have noticed that this delay does not exist if I use the rule below directly in my i3 config instead of using it in i3ipc. 

```
for_window [instance="kitty_btop"] floating enable, resize set 800 800, move position center
```

My question is, (1) is it possible to use `for_window [instance="kitty_btop"]` commands from within i3ipc so that I can set rules from within a script, (2) if this is not possible, how can I modify the function `launch_app`, so that the window opens directly in floating mode and does not have to go through a tiled mode first.

**Comment by medwatt** [comment]
Looks like this forum is dead.

**Comment by okraits** [comment]
Why do you think that this forum is dead?

1. No, because `for_window` is a config directive and not a command.
2. Why don't you define the rule in your config?

**Reply by medwatt** [reply]
For two reasons:

1. It is way easier to manage things and reuse code when configuring things in a proper programming language.

2. I am using i3ipc to improve the use of scratchpads so that when the application is closed, I don't need to restart i3 to respawn the application. However, because I cannot set the names of windows that get moved to the scratchpad in code, I have to wait until the window is created, detect that, and then move it to the scratchpad. On my computer, which is not too slow, I can definitely see the window getting spawned in tiling mode, getting converted to floating mode, and moved to the scratchpad. This causes other windows on the same workspace to momentarily change dimensions back and forth. Defining the rule in i3 config is not going to work because I need to `mark` the window first to distinguish it from other windows so that i3 knows which window to move to the scratchpad.

```python
def toggle_application(self, class_name, cmd):
    con_mark = f"sc_{class_name}"
    # Check if a window with the given mark already exists
    windows = self.i3.get_tree().find_marked(con_mark)
    if windows:
        # If a window exists, show it from the scratchpad
        self.i3.command(f"[con_mark=sc_{class_name}] scratchpad show")
    else:

        self.is_window_in_scratchpad = True

        # If no window exists, launch the application and wait for the window to appear
        self.i3.command(f"exec {cmd}")

        def on_window_event(i3, event):
            if event.container.window_class == class_name:
                # Move the window to the scratchpad and make it floating
                window_id = event.container.window
                create_window = (
                    f"[id={window_id}] floating enable",
                    f"mark sc_{class_name}",
                    "resize set 1000 px 800 px",
                    "move position center",
                    "border pixel 5",
                    "move scratchpad",
                    "scratchpad show",
                )
                self.i3.command(",".join(create_window))

                # Remove the event listener
                self.i3.off(on_window_event)

        self.i3.on(Event.WINDOW_NEW, on_window_event)

```

**Comment by okraits** [comment]
It would make helping you easier if you would precisely describe your usecase and what you are trying to achieve.

**Reply by medwatt** [reply]
I already did that in the first post and expanded on it in the follow-up. To summarize: Using python, I want to be able to set the properties of a window (such as whether it is floating, which workspace it belongs to, etc) **before** the window is spawed. At the moment, I can only set the properties **after** the window is spawned.

**Reply by okraits** [reply]
I think this is due to the way i3 and the ipc interface works. I guess that when a new window appears roughly the following happens:

1. i3 checks wether the window is affected by an `assign` directive in its config.
2. i3 checks wether the window is affected by one or more `for_window` directives in its config.
3. If any of the above is true, i3 puts the window on the defined workspace and executes the command(s) defined in the `for_window` directive(s).
4. If none of the above is true, i3 puts the window on the current workspace in the place the current layout or split container defines for the next new window.
5. i3 creates the `new` window event, notifying all clients using the ipc interface.

This means that even if you define properties for a window to be set with the ipc interface, i3 will still apply the "internal" default properties for that window (defined in its config) first. At least it seems to me that the "internal" properties defined in the config have a higher priority. After all that's the purpose of the config.

**Reply by medwatt** [reply]
@okraits, thanks for your answer, but I'm actually looking for solutions, which, if I am to understand from your comment, seems not to exist.
